src/main/java/hudson/tasks/junit/History.java

```diff
@@ -64,49 +76,473 @@ public TestObject getTestObject() {
+    private LinesChartModel createTestResultTrend(int start, int end, ChartModelConfiguration chartModelConfiguration) {
         TestResultImpl pluggableStorage = getPluggableStorage();
         if (pluggableStorage != null) {
             return new TestResultTrendChart().create(pluggableStorage.getTrendTestResultSummary());
         }
+        return new TestResultTrendChart().createFromTestObject(createBuildHistory(testObject, start, end), chartModelConfiguration);
+    }
+
+    private ObjectNode computeDurationTrendJson(List<HistoryTestResultSummary> history) {
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectNode root = mapper.createObjectNode();
+        ArrayNode domainAxisLabels = mapper.createArrayNode();
+        ArrayNode series = mapper.createArrayNode();
+        ObjectNode durationSeries = mapper.createObjectNode();
+        series.add(durationSeries);
+        durationSeries.put("type", "line");
+        durationSeries.put("symbol", "circle");
+        durationSeries.put("symbolSize", "6");
+        durationSeries.put("sampling", "lttb");
+        ArrayNode durationData = mapper.createArrayNode();
+        durationSeries.set("data", durationData);
+        ObjectNode durationStyle = mapper.createObjectNode();
+        durationSeries.set("itemStyle", durationStyle);
+        durationStyle.put("color", "rgba(160, 173, 177, 0.6)");
+        ObjectNode durationAreaStyle = mapper.createObjectNode();
+        durationSeries.set("areaStyle", durationAreaStyle);
+        durationAreaStyle.put("normal", true);
+        ObjectNode durationMarkLine = mapper.createObjectNode();
+        durationSeries.set("markLine", durationMarkLine);
+        ArrayNode durationMarkData = mapper.createArrayNode();
+        durationMarkLine.set("data", durationMarkData);
+        ObjectNode durationAvgMark = mapper.createObjectNode();
+        ObjectNode hideLabel = mapper.createObjectNode();
+        hideLabel.put("show", false);
+        ObjectNode dashLineStyle = mapper.createObjectNode();
+        dashLineStyle.put("dashOffset", 50);
+        dashLineStyle.put("color", "rgba(128, 128, 128, 0.1)");
+        ArrayNode lightDashType = mapper.createArrayNode();
+        lightDashType.add(5);
+        lightDashType.add(10);        
+        dashLineStyle.set("type", lightDashType);
+        durationAvgMark.put("type", "average");
+        durationAvgMark.put("name", "Avg");
+        durationAvgMark.set("label", hideLabel);
+        durationAvgMark.set("lineStyle", dashLineStyle);
+        durationMarkData.add(durationAvgMark);
+
+        float maxDuration = (float)0.0;
+        for (HistoryTestResultSummary h : history) {
+            hudson.tasks.test.TestResult to = h.getResultInRun();
+            if (maxDuration < to.getDuration()) {
+                maxDuration = to.getDuration();
+            }
+        }
+        ObjectNode yAxis = mapper.createObjectNode();
+        double mul = 1.0;
+        double roundMul = 1000.0;
```

Comment:
Rounding is looking better, 2 issues I'm seeing still though:

1. trend rounding seems longer than the duration one:
<img width="231" alt="image" src="https://github.com/user-attachments/assets/e245a7e6-79f3-4928-bac8-3af9a3fdf6ad">

<img width="964" alt="image" src="https://github.com/user-attachments/assets/4c0bdd8c-3ded-4c5c-9afb-a4666cd6c5e8">

2. I don't think we need any decimal places when using milliseconds, see e.g the table which doesn't show it: 

<img width="539" alt="image" src="https://github.com/user-attachments/assets/6f52a66e-19b2-43eb-9d37-26b4bfb71b9d">

Reply:
I did some improvements.